### PR TITLE
Canny TBB Optimization

### PR DESCRIPTION
**Pull request by oresths**
Generally a good speedup.

I did some tests in a quad core:

Measured 40-80% speedup compared to IPP.
Speedup compared to the pure serial code 2.4x

IPP cpu utilization was 100%
TBB cpu utilization was 265%

So IPP is more efficient, but TBB is faster. Also non-x86 platforms will benefit from TBB.

check_regression=*anny*

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
can not universal `cv::parallel_for` be used here?

**Comment by oresths** [comment]
As the iterations for maxima supppresion / thresholding are not independent (because of the ring buffer, the check to the previous row before pushing a "sure" edge to the stack to save calculations and edge tracking ) I think cv::parallel_for wouldn't work.

As for the Sobel step, propably a good solution would be to parallelize FilterEngine::proceed , but there is also a ring buffer there.
But even if we did that, we wouldn't have the option to use ippSobel to calculate the image slices.

**Comment by vpisarev** [comment]
@oresths, thank you for the contribution!

I have some doubts about the patch though. First of all, as @ilya-lavrenov said, you are using TBB, which we replaced in 99% of cases with the universal parallel_for_. On iOS/OSX and on Windows TBB is not used by default (on iOS it can not be used at all), so it limits the usability of the patch.

Secondly, I have not studied it in details, but from what I see, the edge tracing part has been also parallelized, and since you do not use any atomic operations to read and write the map, there is very big chance of data races in my opinion.

In other words, we can not accept the patch as-is. If you modify the patch to parallelize only the first part of the algorithm, before the edge tracing, which can be done using cv::parallel_for_, then there will be no data races and the threading will be available on all the platforms. Even though the speed may decrease a bit, the algorithm correctness is more important.

**Comment by oresths** [comment]
Hi @vpisarev, thanks for checking the patch.

About OSX and windows not using TBB by default I need some clarification. Do you mean is not used by default because OpenCV doesn't install TBB by default or because in the canny source file I let IPP version first ?
If it is the first case, because IPP is prioritized (I think it should be prioritized because it is more efficient), to use TBB version you should undefine IPP code so TBB can be used. But if you can do that, you know you should install TBB first.
If it is the second case, I can change that easily.
For any case I think is better for the user to decide what's best for their project.

Before I go into further explanation it would help to split current canny implementation into 4 steps (different from the theoretical ones). 
First step is sobel, 
second step is magnitude calculation,
third step is thresholding and
fourth step edge linking.

So, here is an example test case; I used this image https://github.com/oresths/opencv-tests/blob/master/people.jpg , all tests are performed in an Odroid XU running Ubuntu.
With low_thres=100 and high=150
Step1 – 132ms
Step2 – 35ms
Step3 – 83ms
Step4 – 24ms
Time to allocate an image size integer matrix - 64ms 

So for the first step, which is usually the slowest, FilterEngine::proceed should be parallelized. But the ring buffer there prevents the use of parallel_for, because iterations are not independent. Moreover my tests with NEON filters showed that the cost of allocating an integer intermediate buffer would require more time than the filters themselves. So, although this would be the optimal solution, as many other algorithms will benefit, I think proceed function should be rewritten for parallelization, which is a complicated task and probably will not happen any time soon.
Also, the way I implement it users can choose between ippSobel (for x86 platforms) and SSE / NEON, if they so desire. 

Steps 2 and 3, are the steps we could use parallel_for by removing the ring buffer, but with the added cost of allocating an image size integer matrix for magnitudes (64ms – huge performance hit). Also step 2 is fast anyway, so the impact in performance is very small. Step 3 is threshold dependent. I think that with parallel_for the parallel version could be slower than the serial in some cases. 
Instead of removing the ring buffer I would suggest that we make ring buffers for dx and dy sobel results, because allocation for each destination matrix costs 32 ms (another 64ms performance hit !)

For steps 3 and 4 I first tried atomic operations to protect a common stack, but performance boost was disappointing. So I had to find a way to remove locks (which is a good practice anyway when parallelizing)
 I realized that each stripe could be completely independent from other stripes. What I do is that when a task pops a value that is close to the border, I stop the linking, I send it to a TBB concurrent queue, and continue with the next value in the stack. Then after all threads finish, elements in the concurrent queue are processed serially. This had a considerable performance boost. Also, by design there should be no chance of data races (that goes for all the steps, too).

If you can wait, I will make some other time measurements and post them here later.

Some ideas to improve my implementation:
- If IPPDerivSobel in the future supports borders for non - isolated ROIs, the first step could be simplified significantly.

Some other ideas for optimization would be:
- Include the first theoretical Canny step (smoothing filter) in the canny.cpp to prevent yet another allocation
- If the first theoretical step is a linear filter maybe it is worth it to filter the sobel kernels with the smoothing filter first before filtering the whole image.  A study of computational complexity is required here to choose right kernel sizes for filtering the image, without slowing down. Also SymmColumnSmall32s16s for the general antisymmetric case should be implemented with SSE / NEON and createSeparableLinearFilter should change accordingly to allow for integer intermediate buffer, so as to avoid some floating point arithmetic. (Something like this: https://github.com/oresths/opencv/commits/MixedSobelGauss (only 5x1 NEON version implemented))
- Optimize somehow the final image creation (the step that creates the 0 or 255 canny output image). It costs 27ms and 42ms if it wasn’t preallocated.

To sum it up, before writing anything I checked if it could be done with cv::parallel_for, and I had to reject it because it wouldn’t work / doesn’t worth it. So I decided to use TBB which is very portable (except iOS ? - is there a framework in opencv that supports every platform ?) and a very popular addition to opencv.   So I created  a fast  mutex-free and race free version with only minimal modifications to the serial version (that makes for easier maintenance etc). This implementation is by no way the best solution but someone has to make the first step.
If the practice is to not parallelize at all if it is not possible with parallel_for, please feel free to close the PR.

If you have any suggestions / questions I would be happy to reply.

**Comment by vpisarev** [comment]
@oresths, terribly sorry for delay; I hoped, I can find time to do something from the suggested list (like optimizing out Sobel's), but in the end decided to merge your patch and then refine/optimize Canny further. The technique you use looks great. I still believe it can be done with a parallel_for (and then users on Windows and iOS/OSX will enjoy the acceleration as well), but let's do it later. Thanks again for the contribution! :+1:
